# Floating Point Arithmetic and Errors

## Absolute and Relative Error

Let $\alpha$ be an approximation of $a$. Then

$$err=|a-\alpha| \\  relerr=\frac{err}{|a|}=\frac{|a-\alpha|}{|a|}.$$

In [1]:
# Try α=a:0.01:2a
a=5.0
α=5.1
err=abs(a-α)
relerr=err/abs(a)
α, err, relerr

(5.1, 0.09999999999999964, 0.019999999999999928)

## Floating Point Arithmetic

Useful book on IEEE Floating Point standard:

M. Overton, Numerical Computing with IEEE Floating Point Arithmetic, SIAM Publications, Philadephia, 2001.

Useful article: 

[David Goldberg, What Every Computer Scientist Should Know About Floating-Point Arithmetic](https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html).

### Floating Point Number System

$x$ is a floating point number if it has the form

$$
	x = \pm d \cdot \beta^e \quad \beta \in \{ 2,10 \}.
$$

Base 2 is for general purpose computers, base 10 is for pocket calculators.

$e$ is the exponent and satisfies

$$
	e_{\min} \leq e \leq e_{\max},\quad
	e_{\min} < 0 < e_{\max}
$$

We will assume that arithmetic is in base 2, but will usually give examples in base 10.

Mantissa $d$ has the form

\begin{align*}
	d &= 0.d_1 \dots d_t = d_1 \beta^{-1} + d_2 \beta^{-2}
	+ \dots + d_t \beta^{-t}\\
d_i  &\in \{ 0,1\}\\
	d_1 &= 1 \qquad \mbox{ normalized }   \\
	d_1 &= 0 \qquad \mbox{ unnormalized }   \\
\end{align*}

Standard form for floating point numbers is normalized except at the
bottom of the exponent range.

During input and output numbers are converted from binary to decimal
and back.

Computer arithmetic is standardized, by the IEEE 754 standard
for binary arithmetic.  All but a few modern computers follow this standard.

### Machine Unit

_Machine unit_ (or _machine precision_) the smallest number $\epsilon$ such that $1+\epsilon\neq 1$:

$$
	\epsilon = \max_{\lfloor \log_2 
    \:|x|\rfloor \in
	[e_{\min},e_{\max}]} \frac{|x - fl(x)|}{|x|}  = 2^{-t}
$$

The set

$$
\{ x \colon \lfloor \log_2 \: |x| \rfloor \in [e_{min},e_{max}] \}
$$

is the set of real numbers that are in the normalized range of floating point numbers.
$fl(x)$ is the floating point round of $x$. Thus the _machine unit_ is the maximum relative distance between a real number in the floating point range and the nearest floating point number.

Important examples include

_IEEE Standard Single Precision_ (`Float32`):  $\beta = 2$, $t = 24$,

\begin{align*}
	\epsilon_M  &= 2^{-24} \approx	5.9605 \times 10^{-8}\\
	e_{\min} &= - 126,\quad e_{\max} = 128.
\end{align*}


_IEEE Standard Double Precision_ (`Float 64`): $\beta =2$, $t = 53$,

\begin{align*}
	\epsilon &= 2^{-53} \approx 1.1102 \times 10^{-16}\\
    e_{\min} &= -1022,\quad e_{\max} = 1024
\end{align*}

Let us compute machine unit as the smallest positive floating point number $\epsilon$ such that  $1+\epsilon\neq 1$:

In [2]:
b=1.0
a=2.0
while (b+a)!=b
    a=a/2
    println(a)
end

1.0
0.5
0.25
0.125
0.0625
0.03125
0.015625
0.0078125
0.00390625
0.001953125
0.0009765625
0.00048828125
0.000244140625
0.0001220703125
6.103515625e-5
3.0517578125e-5
1.52587890625e-5
7.62939453125e-6
3.814697265625e-6
1.9073486328125e-6
9.5367431640625e-7
4.76837158203125e-7
2.384185791015625e-7
1.1920928955078125e-7
5.960464477539063e-8
2.9802322387695312e-8
1.4901161193847656e-8
7.450580596923828e-9
3.725290298461914e-9
1.862645149230957e-9
9.313225746154785e-10
4.656612873077393e-10
2.3283064365386963e-10
1.1641532182693481e-10
5.820766091346741e-11
2.9103830456733704e-11
1.4551915228366852e-11
7.275957614183426e-12
3.637978807091713e-12
1.8189894035458565e-12
9.094947017729282e-13
4.547473508864641e-13
2.2737367544323206e-13
1.1368683772161603e-13
5.684341886080802e-14
2.842170943040401e-14
1.4210854715202004e-14
7.105427357601002e-15
3.552713678800501e-15
1.7763568394002505e-15
8.881784197001252e-16
4.440892098500626e-16
2.220446049250313e-16
1.1102230246251565e-16


The MATLAB eps `eps` and the Julia function `eps()` give $2.2204 \times
10^{-16}$, which is the maximum relative spacing between two
floating point numbers. As you can easily deduce, this number is $2\epsilon$.

In [3]:
eps()

2.220446049250313e-16

In [4]:
# What is this?
eps(200.0)

2.842170943040401e-14

Julia, in particular, has a type system where `Float64` type is a sub-type of `AbstractFloat`, which has four sub-types. 
In addition to types `Float64` and `Float32`, there is a type `Float16` which uses only two bytes of computer memory and type `BigFloat` which has a 256-bit mantissa.  

In [5]:
supertype(Float64)

AbstractFloat

In [6]:
subtypes(AbstractFloat)

4-element Array{Any,1}:
 BigFloat
 Float16 
 Float32 
 Float64 

In [7]:
for T in (Float16, Float32, Float64, BigFloat)
    println(eps(T))
end

0.000977
1.1920929e-7
2.220446049250313e-16
1.727233711018888925077270372560079914223200072887256277004740694033718360632485e-77


In [8]:
2^(-10), 2^(-23), 2^(-52), 2^(-255)

(0.0009765625, 1.1920928955078125e-7, 2.220446049250313e-16, 1.727233711018889e-77)

### Basic Floating Point Operations

We begin with the four basic arithmetic operations, addition ($+$),subtraction ($-$),multiplication ($*$), and division ($/$). Suppose $\odot$ is an operation such that

$$
\odot \in \{ + , - , *,/\}.
$$

Then, in floating point arithmetic with machine unit $\epsilon$, it is reasonable to expect that
for any two floating point numbers $x$ and $y$, we have


$$
fl(x\odot y) = (x \odot y)\;(1 + \xi),\quad
|\xi| \leq \epsilon.
$$

For division, we assume $y \neq 0$.
Any IEEE standard computer must follow this rule.  Rounding is one of two limitations that floating point
arithmetic has that real arithmetic does not have. You can quickly conclude from the above rule that as long as all that we do is add numbers of the same sign, multiply, and divide, floating point results will almost always come very close to the corresponding real arithmetic results. The difficulty occurs if we either of $x$ or $y$ is rounded, they have different signs and we add or have the same signs and we subtract. 

That is, suppose we have

$$
\tilde{x}= x(1+\delta_x), \quad \tilde{y} = y(1+\delta_y),
$$

where $x$ and $y$ are the exact real results of some computation and $\tilde{x}$ and $\tilde{y}$ are rounded floating point results with $|\delta_x| |\delta_y| \leq \delta$ for some small $\delta$. Suppose also that $x$ and $y$ have the same sign. If

$$
z=x-y,\quad  \tilde{z} = fl(\tilde{x} -\tilde{y})
$$

then

\begin{align*}
\tilde{z} &=(\tilde{x}-\tilde{y})(1+\xi)= x(1+\delta_x)(1+\xi) -y(1+\delta_y)(1+\xi) 
=x-y + \delta_z,
\end{align*}

where $|\xi| \leq \epsilon$ and

$$
\delta_z = (x-y)\xi + (x\delta_x -y\delta_y)(1+\xi).
$$

The best available bound on $|\delta_z|$ is

\begin{align*}
|\delta_z| &\leq |x-y||\xi| + (|x||\delta_x| + |y||\delta_y|)(1+|\xi|) \\
& \leq |x-y| \epsilon + (|x|+|y|)\,\delta\,(1+\epsilon).
\end{align*}

Thus, 

\begin{align*}
|\delta_z| &= \frac{|\tilde{z}-z|}{|z|}
\leq \epsilon + (1+\epsilon)\,\delta\,\frac{|x|+|y|}{|x-y|}.
\end{align*}

If $|x-y| \ll |x|+|y|$, the effect of the round in the subtraction is not important, but the error from previous computations on $x$ and $y$ can have a huge effect.  
The effect is called _propagation_ or _catastrophic cancellation_. It can dramatically change the result of a compuation! We will see this issue with some examples later in this lecture.

Rounding is the first important limitation of floating point arithmetic. A second limitation is the number range.

### Number Ranges

Floating point arithmetic has a largest and smallest computer number. First, the largest one.

__Largest Computer Number__

In base $2$, with a $t$ bit mantissa, the largest computer number is

$$
	\Omega = (1 - 2^{-t}) \cdot 2^{e_{\max}}
$$

When numbers exceed $\Omega$ (we say there is an _overflow_), they are stored as `Inf` ($\infty$) or `-Inf` ($-\infty$).


_IEEE Standard Single Precision_ (`Float32`):

$$
e_{\max} = 128,  \quad \Omega = 3.4028\approx 10^{38}
$$


_IEEE Standard Double Precision_ (`Float64`):

$$
e_{\max} = 1024, \quad \Omega = 1.79777 \times 10^{308}
$$

MATLAB naredba `realmax` i Julia funkcija `floatmax()` prikazuju ovaj broj.

__Smallest Computer Number__

The definition of the smallest computer number is somewhat more complex.

The smallest computer number is given by

$$
\omega = 2^{1-t} 2^{e_{\min}}.
$$

If a computation produces a number smaller in magnitude than $\omega$, it produces what is called an _underflow_, it is set to $0$ or
$-0$.  If the programmer chooses, an underflow can result in an error, but in most computations, underflows are not harmful.



_IEEE Standard Single Precision_ (`Float32`):

$$
\omega = 2^{-23- 126} = 2^{-149} \approx  1.4013 \times 10^{-45}.
$$

In MATLAB, this comes from the command `omega= eps('single')*realmin('single')`.


_IEEE Standard Double Precision _ (`Float64`):

$$
\omega= 2^{-1022-52} = 2^{-1074} \approx  4.9407 \times 10^{-324}
$$

The appropriate MATLAB command to get this value is
`omega = eps*realmin`. The corresponding Julia command is `floatmin()*eps()`.


_Important and Subtle Point_ 

Numbers at the bottom of the exponent range are not normalized. 
MATLAB function `realmin` yields

$$
\omega_{useful} \approx 2.2251 \times 10^{-308}.
$$

Some people call this the smallest USEFUL floating point number since

$$
1/\omega_{useful} \leq \Omega,
$$

and $\omega_{useful}$ is normalized.

Smallest floating point number, $\omega$, has the form

$$
0.0 \cdots 01 \times 2^{e_{\min}} \quad \cdots\quad
\mbox{gradual underflow}
$$

Before the IEEE standard most computers had the smallest
floating point number as

$$
	0.10 \cdots 0 \times 2^{e_{\min}} \qquad \cdots
	\mbox{ normalized}
$$

Earlier computers, (pre-1985) set numbers below this smallest 'useful' floating point number to zero. This change was one of the more controversial features of the IEEE standard.
Here is an example (in decimal arithmetic) which shows why this feature made it into the IEEE standard.


__Example__ $\beta = 10$, $-5 \leq e \leq 5$

\begin{eqnarray*}
	x & = 0.1957 \times 10^{-5}   \\
	y & = 0.1942 \times 10^{-5}
\end{eqnarray*}

Do  $fl(x - y)$. What happens?

$$
0.1957 \times 10^{-5}-0.1942 \times 10^{-5}  =0.0015 \times 10^{-5}
$$

The pre-1985 philosophy was to set $fl(x - y)=0$.

Gradual underflow stores $fl(x - y)=0.0015 \times 10^{-5}$.

Gradual underflow guarantees that for any two floating point numbers $x$ and $y$, 

$$
fl(x - y) = 0 \mbox{ if and only if } x = y.
$$

In [9]:
for T in (Float16, Float32, Float64, BigFloat)
    println((floatmin(T),floatmax(T)))
end

(Float16(6.104e-5), Float16(6.55e4))
(1.1754944f-38, 3.4028235f38)
(2.2250738585072014e-308, 1.7976931348623157e308)
(8.50969131174083613912978790962048280567755996982969624908264897850135431080301e-1388255822130839284, 5.875653789111587590936911998878442589938516392745498308333779606469323584389875e+1388255822130839282)


In [10]:
1/floatmin(),floatmax()

(4.49423283715579e307, 1.7976931348623157e308)

In [11]:
for T in (Float16, Float32, Float64)
    println((floatmin(T)*eps(T)))
end

6.0e-8
1.0e-45
5.0e-324


###  Special Quantities $0$, $-0$, `Inf` i `NaN`

Zero has a sign:

In [12]:
a=1.0
b=0.0
c=-b
c,b==c

(-0.0, true)

In [13]:
d=a/b
e=a/c
d==e, 1/d==1/e

(false, true)

In [14]:
b/c

NaN

In [15]:
bitstring(0)

"0000000000000000000000000000000000000000000000000000000000000000"

In [16]:
bitstring(1)

"0000000000000000000000000000000000000000000000000000000000000001"

In [17]:
bitstring(0.0)

"0000000000000000000000000000000000000000000000000000000000000000"

In [18]:
bitstring(-0.0)

"1000000000000000000000000000000000000000000000000000000000000000"

In [19]:
bitstring(1.0)

"0011111111110000000000000000000000000000000000000000000000000000"

__Question.__ Explain the above binary representations. 

### Example

__Using difference of squares__

Compute

$$
f(x) = \sqrt{1 + x^2} - 1, \quad \mbox{$x$ near zero}.
$$

Using this formula in standard double precision yields
$f(10^{-12}) = 0$.

In [20]:
f(x)=sqrt(1+x^2)-1
x=1e-6
for k=1:4
    println((x,f(x)))
    x=x/10
end

(1.0e-6, 5.000444502911705e-13)
(1.0e-7, 4.884981308350689e-15)
(1.0e-8, 0.0)
(1.0e-9, 0.0)


An old "difference of two squares" trick gives

\begin{eqnarray*}
f(x) & \equiv (\sqrt{1 + x^2} - 1) \left( \frac{\sqrt{1 + x^2} + 1}{\sqrt{1 + x^2} + 1}\right) \\
& = \frac{x^2}{\sqrt{1+x^2} + 1}\equiv f_1(x),
\end{eqnarray*}

that is, $f_1(10^{-12}) = 0.5 \cdot 10^{-24}$. This answer is as accurate as we can expect to get in this precision.

In [21]:
f₁(x)=x^2/(1+sqrt(1+x^2))
x=1e-6
for k=1:10
    println((x, f₁(x)))
    x=x/10
end

(1.0e-6, 4.99999999999875e-13)
(1.0e-7, 4.999999999999987e-15)
(1.0e-8, 5.0000000000000005e-17)
(1.0e-9, 5.0e-19)
(1.0e-10, 5.0000000000000005e-21)
(1.0000000000000001e-11, 5.000000000000001e-23)
(1.0000000000000002e-12, 5.000000000000001e-25)
(1.0000000000000002e-13, 5.0000000000000016e-27)
(1.0000000000000002e-14, 5.0000000000000015e-29)
(1.0e-15, 5.0e-31)


__Quadratic Equation__

In exact arithmetic, the quadratic equation

$$ ax^2 + bx+c=0$$

has roots

\begin{align*}
x_1&=\frac{-b-\sqrt{b^2-4ac}}{2a} \\
x_2&\equiv\frac{-b+\sqrt{b^2-4ac}}{2a}= \frac{-b+\sqrt{b^2-4ac}}{2a}\cdot \frac{-b-\sqrt{b^2-4ac}}{-b-\sqrt{b^2-4ac}}
\\ &= \frac{2c}{-b-\sqrt{b^2-4ac}}\equiv x_3.
\end{align*}

In [22]:
a=2.0
b=123456789.0
c=4.0

x₁=(-b-sqrt(b*b-4*a*c))/(2.0*a)
x₂=(-b+sqrt(b*b-4*a*c))/(2.0*a)
x₃=(2*c)/(-b-sqrt(b*b-4*a*c))
x₁,x₂,x₃

(-6.172839449999997e7, -3.3527612686157227e-8, -3.240000029484002e-8)

Check with `BigFloat`:

In [23]:
a=BigFloat(a)
b=BigFloat(b)
c=BigFloat(c)
x₂=(-b+sqrt(b*b-4*a*c))/(2.0*a)

-3.240000029484001968915648868258452417675753633383540995167795107129921671968718e-08

__Tangent and sine__

In [24]:
x=1e-10
tan(x)-sin(x)

0.0

However, the trigonometric identities give

\begin{align*}
\tan x - \sin x & = \tan x (1 - \cos x ) 
= \tan x (1-\cos x)\frac{1+\cos x}{1+\cos x}\\ & = \tan x \frac{1-\cos^2 x}{1+\cos x} \\
&= \tan x \sin^2 x \frac{1}{1+\cos x},
\end{align*}

and the Taylor formula gives

\begin{align*}
\tan x &= x + \frac{x^3}{3} + \frac{2x^5}{15} + O(x^7) \\
\sin x &= x -\frac{x^3}{6} + \frac{x^5}{120}+O(x^7) \\
\tan x - \sin x &= \frac{x^3}{2} + \frac{7x^5}{120} +O(x^7)
\end{align*}

Both formulas give accurate result:

In [25]:
tan(x)*sin(x)^2/(1+cos(x)), x^3/2+7*x^5/120

(5.0e-31, 5.0e-31)

__Absolute Value of a Complex Number__

In order to avoid undeflow or overflow, instead of the standard formula 

$$
|z|=|x+iy|=\sqrt{x^2+y^2}
$$

we need to use the following formulas (explain!):

$$
M = \max \{ |x|,|y|\}, \quad m = \min \{ |x|,|y| \}, \quad r = \frac{m}{M}, \quad 
|z| = M \sqrt{1+r^2}.
$$

In [26]:
z=2e-170+3e-175*im

2.0e-170 + 3.0e-175im

In [27]:
sqrt(real(z)^2+imag(z)^2), abs(z)

(0.0, 2.000000000225e-170)